### PR TITLE
Bug 1201108 - Incremental server bookmark mirrorer.

### PR DESCRIPTION
**Pull request by rnewman**
WIP. Builds on #1055.

### CODE REVIEW

**Comment by fluffyemily** [comment]
when trying to run tests, I'm getting a build error on TestSQLiteBookmarks: testBookmarks() lns 25 & 38

**Comment by rnewman** [comment]
Tests fixed!

**Comment by fluffyemily** [comment]
:+1:

File: SyncTests/DownloadTests.swift [hunk 1]
```diff
@@ -81,4 +81,50 @@ class DownloadTests: XCTestCase {
...
+        let prefs = MockProfilePrefs()
+
+        let batcher = BatchingDownloader(collectionClient: bookmarksClient, basePrefs: prefs, collection: "clients")
+
+        let ic1 = InfoCollections(collections: ["clients": ts1])
+        let fetch1 = batcher.go(ic1, limit: 1).value
+        XCTAssertEqual(fetch1.successValue, DownloadEndState.Complete)
+        let records1 = batcher.retrieve()
+        XCTAssertEqual(1, records1.count)
+        XCTAssertEqual(guid1, records1[0].id)
```
Comment (fluffyemily) [inline SyncTests/DownloadTests.swift]: Getting an EXC_BAD_INSTRUCTION here when executing tests. `records1` has 0 elements.

File: SyncTests/RecordTests.swift [hunk 2]
```diff
@@ -228,16 +231,35 @@ class RecordTests: XCTestCase {
...
+        }
+
+        XCTAssertEqual(6, item.type.rawValue)
+        XCTAssertEqual("ShCZLGEFQMam", item.guid)
+
+        let places = JSON.parse("{\"id\":\"places\",\"type\":\"folder\",\"title\":\"\",\"description\":null,\"children\":[\"menu________\",\"toolbar_____\",\"tags________\",\"unfiled_____\",\"jKnyPDrBQSDg\",\"T6XK5oJMU8ih\"],\"parentid\":\"2hYxKgBwvkEH\"}")
+        let p = BookmarkType.payloadFromJSON(places)
+        XCTAssertTrue(p is FolderPayload)
+        XCTAssertTrue(p is MirrorItemable)
+        XCTAssertEqual(BookmarkRoots.RootGUID, p!.id)
```
Comment (fluffyemily) [inline SyncTests/RecordTests.swift]: test failure:`(Optional( "root_____")) is not equal to (Optional("places"))`

**Comment by rnewman** [comment]
This was landed and re-enabled in 71449b8.

**Pull request closed without merging**